### ISSUE TITLE
[vim9script] unexpected error accessing items imported through symbolic linked directory

### ISSUE THREAD

**Issue by errael** [vim/vim#14536]
### Steps to reproduce

**NOTE:** for better example, with additional debug output, see https://github.com/vim/vim/issues/14536#issuecomment-2054136374

At the entry of `as_dir/f1.vim` there is

```
vim9script
echomsg 'EXECUTING f1'
echomsg expand('<stack>')
```

There's the following output, slightly edited for clarity.
The important point is that **the import file f1 is loaded twice**.
Which, IIUC, is impossible; `vim` shouldn't load the same import file twice.

```
lib_dir: /home/err/bugs/vim/import-symlink/syml_dir
EXECUTING f1
:source buffer=1[22]
    function <SNR>39_Fmain[1]
    script /home/err/bugs/vim/import-symlink/as_dir/f1.vim[3]
EXECUTING f1
:source buffer=1[22]
    function <SNR>39_Fmain[1]
    script /home/err/bugs/vim/import-symlink/as_dir/f1.vim[3]

Error detected while processing :source buffer=1[22]..function <SNR>39_Fmain:
line    1:
E117: Unknown function: <SNR>40_Ff1
```

Symlinks are involved in the problem.
Note in the source three ways to workaround the problem.
1. `import lib_dir .. '/f1.vim'`
2. `const Ff1 = f1.Ff1`
3. import from the directory, not `syml_dir`: `.. '/as_dir'`

(1) and (2) loads the import before the function `Fmain()` executes. So it
seems like the problem only occurs during a function invocation when it needs
to do an autoload and there is a symlink involved. My hand waving guess is
that when there's a synlink involved it's using the wrong thing to check if
the file has already been loaded. (3) avoids the symlink.

Set up the following files in this structure:
```
.
├── as_dir
│   └── f1.vim
├── main.vim
└── syml_dir -> as_dir
```

**main.vim**
```
vim9script

export var lib_dir: string = fnamemodify(
    getscriptinfo(
        {sid: str2nr(matchstr(expand('<SID>'), '\v\d+'))}
    )[0].name, ':p:h') .. '/syml_dir'

echomsg 'lib_dir:' lib_dir


# REMOVE autoload FROM FOLLOWING LINE AND IT WORKS
import autoload lib_dir .. '/f1.vim'

# UNCOMMENT THE FOLLOWING LINE AND IT WORKS
#const Ff1 = f1.Ff1

def Fmain()
    var xxx = f1.Ff1()
enddef
Fmain()
```

**syml_dir/f1.vim, as_dir/f1.vim**
```
vim9script
echomsg 'EXECUTING f1'
echomsg expand('<stack>')

export const v1val = 1

export def Ff1(): number
    return v1val
enddef
```

### Expected behaviour

No error.

I'm going to look at tracking this down, but I've never looked an imports before...

Any hints?

### Version of Vim

9.1.301

### Environment

ubuntu/gtk

**Comment by errael** [comment]
It occurs to me this is probably a script-level versus `:def` issue.

Just need to find some time...

**Comment by errael** [comment]
Some additional debug output.
Each file outputs `expand('<SID>')`. Disassemble Fmain.

The additional output shows the source of the error. Seems to be the compiler.
The compiler loads the real file, but compiles references to the symlink.

**main.vim**
```
vim9script

export var lib_dir: string = fnamemodify(
    getscriptinfo(
        {sid: str2nr(matchstr(expand('<SID>'), '\v\d+'))}
    )[0].name, ':p:h') .. '/syml_dir'

echomsg expand('<SID>') 'lib_dir:' lib_dir

import autoload lib_dir .. '/f1.vim'

# UNCOMMENT THE FOLLOWING LINE AND IT WORKS
#const Ff1 = f1.Ff1

def Fmain()
    var xxx = f1.Ff1()
enddef

disa Fmain
echo 'After disassemble'

Fmain()
```

**syml_dir/f1.vim,as_dir/f1.vim**
```
vim9script
echomsg 'EXECUTING f1'
echomsg expand('<SID>') expand('<stack>')

export const v1val = 1

export def Ff1(): number
    return v1val
enddef
```

**=== ERROR ===**

Note that the compiler `SOURCE`s the symlink and uses "<80><fd>R40_Ff1",
which is the symlink. But `vim` loads the real file, so an error.
```
    <SNR>41_ .../as_dir/...
```


```
<SNR>39_ lib_dir: /home/err/bugs/vim/import-symlink/ex2/syml_dir
<SNR>39_Fmain
    var xxx = f1.Ff1()
   0 SOURCE /home/err/bugs/vim/import-symlink/ex2/syml_dir/f1.vim
   1 PUSHFUNC "<80><fd>R40_Ff1"
   2 PCALL top (argc 0)
   3 PCALL end
   4 STORE $0
   5 RETURN void
After disassemble
EXECUTING f1
<SNR>41_ :source buffer=1[22]..function <SNR>39_Fmain[1]..script /home/err/bugs/vim/import-symlink/e
x2/as_dir/f1.vim[3]
EXECUTING f1
<SNR>41_ :source buffer=1[22]..function <SNR>39_Fmain[1]..script /home/err/bugs/vim/import-symlink/e
x2/as_dir/f1.vim[3]
Error detected while processing :source buffer=1[22]..function <SNR>39_Fmain:
line    1:
E117: Unknown function: <SNR>40_Ff1
```

**=== OK ===**

With `const Ff1 = f1.Ff1`; then the import happens before Fmain is compiled.

Note that the compiler uses "<80><fd>R41_Ff1".

```
<SNR>39_ lib_dir: /home/err/bugs/vim/import-symlink/ex2/syml_dir
EXECUTING f1
<SNR>41_ :source buffer=1[13]../home/err/bugs/vim/import-symlink/ex2/as_dir/f1.vim[3]
<SNR>39_Fmain
    var xxx = f1.Ff1()
   0 PUSHFUNC "<80><fd>R41_Ff1"
   1 PCALL top (argc 0)
   2 PCALL end
   3 STORE $0
   4 RETURN void
After disassemble
```

**Comment by errael** [comment]
@yegappan Below is a patch that fixes this example.

Compare the following disassembly. Note that with the patch the real
file is used for `SOURCE` and loading. The import logic has lots of
moving parts, so not sure if the patch is complete (or that it doesn't
create other problems).

The only additional overhead is checking at import time for a symlink.
It might be possible to do what's needed during `compile_load_scriptvar()`
after it discovers it's dealing with an import (and so skip the check
during import and avoid any additional runtime overhead), but that feels
more complex (but I don't know if it really is).


```
<SNR>39_ lib_dir: /home/err/bugs/vim/import-symlink/ex2/syml_dir
<SNR>39_Fmain
    var xxx = f1.Ff1()
   0 SOURCE /home/err/bugs/vim/import-symlink/ex2/as_dir/f1.vim
   1 PUSHFUNC "<80><fd>R41_Ff1"
   2 PCALL top (argc 0)
   3 PCALL end
   4 STORE $0
   5 RETURN void
After disassemble
EXECUTING f1
<SNR>41_ :source buffer=1[22]..function <SNR>39_Fmain[1]..script /home/err/bugs/vim/import-symlink/e
x2/as_dir/f1.vim[3]
```


If an import is through a symbolic link then at import time
create a script_items entry for the real file. At compilation,
compile_load_scriptvar(), adjust to use the real file for
"ISN_SOURCE" and var loading.

```diff
diff --git a/src/scriptfile.c b/src/scriptfile.c
--- a/src/scriptfile.c
+++ b/src/scriptfile.c
@@ -402,6 +402,19 @@
        si->sn_name = vim_strsave(fname);
        si->sn_state = SN_STATE_NOT_LOADED;
     }
+
+    // If fname is a symbolic link, create an script_item for the real file.
+    char_u *real_fname = fix_fname(fname);
+    if (STRCMP(real_fname, fname) != 0)
+    {
+       int real_sid = find_script_by_name(real_fname);
+       int error2 = OK;
+       if (real_sid < 0)
+           real_sid = get_new_scriptitem_for_fname(&error2, real_fname);
+       if (error2 == OK)
+           SCRIPT_ITEM(sid)->sn_sourced_sid = real_sid;
+    }
+    vim_free(real_fname);
     return sid;
 }
 
diff --git a/src/vim9expr.c b/src/vim9expr.c
--- a/src/vim9expr.c
+++ b/src/vim9expr.c
@@ -546,6 +546,8 @@
        int     done = FALSE;
        int     res = OK;
 
+       import_check_sourced_sid(&import->imp_sid);
+
        // Need to lookup the member.
        if (*p != '.')
        {
diff --git a/src/vim9script.c b/src/vim9script.c
--- a/src/vim9script.c
+++ b/src/vim9script.c
@@ -387,6 +387,8 @@
 
        si = SCRIPT_ITEM(*sid);
        si->sn_import_autoload = TRUE;
+       if (si->sn_sourced_sid > 0)
+           SCRIPT_ITEM(si->sn_sourced_sid)->sn_import_autoload = TRUE;
 
        if (si->sn_autoload_prefix == NULL)
            si->sn_autoload_prefix = get_autoload_prefix(si);
```

**Comment by errael** [comment]
@yegappan Satisfying my performance obsession...
Looks like it was straightforward to do everything
while compiling. So the following has no runtime
performance implications. Other caveats still apply.

```
If an import is through a symbolic link then at first encounter
during compilation create a script_items entry for the real file;
and adjust to use the real file for "ISN_SOURCE" and var loading.
```

```diff
diff --git a/src/proto/scriptfile.pro b/src/proto/scriptfile.pro
--- a/src/proto/scriptfile.pro
+++ b/src/proto/scriptfile.pro
@@ -9,6 +9,7 @@
 void set_context_in_runtime_cmd(expand_T *xp, char_u *arg);
 int find_script_by_name(char_u *name);
 int get_new_scriptitem_for_fname(int *error, char_u *fname);
+void check_script_symlink(int sid);
 int do_in_path(char_u *path, char *prefix, char_u *name, int flags, void (*callback)(char_u *fname, void *ck), void *cookie);
 int do_in_runtimepath(char_u *name, int flags, void (*callback)(char_u *fname, void *ck), void *cookie);
 int source_runtime(char_u *name, int flags);
diff --git a/src/scriptfile.c b/src/scriptfile.c
--- a/src/scriptfile.c
+++ b/src/scriptfile.c
@@ -405,6 +405,42 @@
     return sid;
 }
 
+/*
+ * If the script for "sid" is a symlink and "sn_source_sid" is not set
+ * then initialize it. A new script_item is created if needed.
+ */
+    void
+check_script_symlink(int sid)
+{
+    scriptitem_T *si = SCRIPT_ITEM(sid);
+    if (si->sn_sourced_sid > 0)
+	return;
+
+    // If fname is a symbolic link, create an script_item for the real file.
+
+    char_u *real_fname = fix_fname(si->sn_name);
+    if (real_fname != NULL && STRCMP(real_fname, si->sn_name) != 0)
+    {
+	int real_sid = find_script_by_name(real_fname);
+	int error2 = OK;
+	int new_sid = FALSE;
+	if (real_sid < 0)
+	{
+	    real_sid = get_new_scriptitem_for_fname(&error2, real_fname);
+	    new_sid = TRUE;
+	}
+	if (error2 == OK)
+	{
+	    si = SCRIPT_ITEM(sid);
+	    si->sn_sourced_sid = real_sid;
+	    if (new_sid)
+		SCRIPT_ITEM(real_sid)->sn_import_autoload
+			  = si->sn_import_autoload;
+	}
+    }
+    vim_free(real_fname);
+}
+
     static void
 find_script_callback(char_u *fname, void *cookie)
 {
diff --git a/src/vim9expr.c b/src/vim9expr.c
--- a/src/vim9expr.c
+++ b/src/vim9expr.c
@@ -546,6 +546,9 @@
 	int	done = FALSE;
 	int	res = OK;
 
+	check_script_symlink(import->imp_sid);
+	import_check_sourced_sid(&import->imp_sid);
+
 	// Need to lookup the member.
 	if (*p != '.')
 	{
```

**Comment by errael** [comment]
Closed by #14565
